tests/unit/test_characterProcessing.py

```diff
@@ -0,0 +1,40 @@
+# A part of NonVisual Desktop Access (NVDA)
+# This file is covered by the GNU General Public License.
+# See the file COPYING for more details.
+# Copyright (C) 2020 NV Access Limited
+
+"""Unit tests for the characterProcessing module.
+"""
+
+import unittest
+import re
+from characterProcessing import SYMLVL_ALL
+from characterProcessing import SpeechSymbolProcessor
+from characterProcessing import processSpeechSymbols as process
+
+
+class TestComplex(unittest.TestCase):
+	"""Test the complex symbols rules.
+	"""
+
+	def _replace(self, m):
+		"""Replacement regexp callback for group replacement, see test_replacement"""
+		group = m.lastgroup
+		if group == 'bar':
+			return SpeechSymbolProcessor._replaceGroups(self, m, r"\\ \1 \2 \3 \\")
+		return m.group()
+
+	def test_replacement(self):
+		"""Test group replacement function"""

```

Comment:
While this may test most of the paths (which is quite hard to determine at a glance), and is a good "happy path" test. I would like to see smaller examples that shows each of the paths. There are at least 6 paths through the `_replaceGroups` function, and many more combinations due to the internal loop. Please add an independent test for each path it can take. 

These tests act as documentation for what the expectations are for the function, so an unfamiliar developer can easily understand the different things the function can do.

Please also test the error case.

Reply:
Ok, I have separated out the tests.